### PR TITLE
SDIT-1829 Refactor to prepare for handling multiple prisoners

### PR DESCRIPTION
**Pull request by mikehalmamoj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/prisonperson/PrisonPersonNomisApiService.kt [hunk 1]
```diff
@@ -20,7 +20,7 @@ class PrisonPersonNomisApiService(@Qualifier("nomisApiWebClient") private val we
 
   suspend fun getReconciliation(offenderNo: String): PrisonPersonReconciliationResponse? =
     webClient.get()
-      .uri("/prisoners/{offenderNo}/physical-attributes/reconciliation", offenderNo)
+      .uri("/prisoners/{offenderNo}/prison-person/reconciliation", offenderNo)
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/prisonperson/PrisonPersonNomisApiService.kt:23]: Fix wrong DPS URL!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/prisonperson/PrisonPersonReconService.kt [hunk 2]
```diff
@@ -9,28 +9,32 @@ import uk.gov.justice.digital.hmpps.prisonertonomisupdate.config.trackEvent
 import uk.gov.justice.digital.hmpps.prisonertonomisupdate.nomissync.model.PrisonPersonReconciliationResponse
 import uk.gov.justice.digital.hmpps.prisonertonomisupdate.prisonperson.model.PhysicalAttributesDto
 import uk.gov.justice.digital.hmpps.prisonertonomisupdate.services.awaitBoth
+import uk.gov.justice.digital.hmpps.prisonertonomisupdate.services.doApiCallWithRetries
 
 @Service
 class PrisonPersonReconService(
   private val dpsApi: PrisonPersonDpsApiService,
-  private val nomisApi: PrisonPersonNomisApiService,
+  private val prisonPersonNomisApi: PrisonPersonNomisApiService,
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/prisonperson/PrisonPersonReconService.kt:17]: Paren NomisApi will be needed for getting active prisoner IDs.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/prisonperson/PrisonPersonReconService.kt [hunk 3]
```diff
@@ -9,28 +9,32 @@ import uk.gov.justice.digital.hmpps.prisonertonomisupdate.config.trackEvent
...
   private val telemetryClient: TelemetryClient,
 ) {
 
-  suspend fun checkPrisoner(offenderNo: String) = runCatching {
+  suspend fun checkPrisoner(offenderNo: String): String? = runCatching {
     val (nomisPrisoner, dpsPrisoner) = withContext(Dispatchers.Unconfined) {
-      async { nomisApi.getReconciliation(offenderNo) } to
-        async { dpsApi.getPhysicalAttributes(offenderNo) }
+      async { doApiCallWithRetries { prisonPersonNomisApi.getReconciliation(offenderNo) } } to
+        async { doApiCallWithRetries { dpsApi.getPhysicalAttributes(offenderNo) } }
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/prisonperson/PrisonPersonReconService.kt:24]: Add in retries

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/prisonperson/PrisonPersonReconService.kt [hunk 4]
```diff
@@ -9,28 +9,32 @@ import uk.gov.justice.digital.hmpps.prisonertonomisupdate.config.trackEvent
...
     }.awaitBoth()
 
-    findDifferences(offenderNo, nomisPrisoner, dpsPrisoner)
-      .takeIf { it.isNotEmpty() }
-      ?.also {
-        telemetryClient.trackEvent("prison-person-reconciliation-report-failed", it)
+    val differences = findDifferences(offenderNo, nomisPrisoner, dpsPrisoner)
+    return if (differences.isNotEmpty()) {
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/prisonperson/PrisonPersonReconService.kt:28]: Return the failed prisoner number for reporting

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/prisonperson/PrisonPersonReconService.kt [hunk 5]
```diff
@@ -9,28 +9,32 @@ import uk.gov.justice.digital.hmpps.prisonertonomisupdate.config.trackEvent
...
+      offenderNo.also {
+        telemetryClient.trackEvent("prison-person-reconciliation-prisoner-failed", differences)
       }
+    } else {
+      null
+    }
   }.onFailure {
-    telemetryClient.trackEvent("prison-person-reconciliation-report-error", mapOf("offenderNo" to offenderNo, "error" to (it.message ?: "unknown error")))
-  }
+    telemetryClient.trackEvent("prison-person-reconciliation-prisoner-error", mapOf("offenderNo" to offenderNo, "error" to (it.message ?: "unknown error")))
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/prisonperson/PrisonPersonReconService.kt:36]: Differentiate between prisoner/report failures and errors

File: src/test/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/prisonperson/PrisonPersonDpsApiMockServer.kt [hunk 6]
```diff
@@ -63,11 +63,12 @@ class PrisonPersonDpsApiMockServer : WireMockServer(WIREMOCK_PORT) {
   }
 
   fun stubGetPhysicalAttributes(
+    offenderNo: String,
     height: Int? = 180,
     weight: Int? = 80,
   ) {
     stubFor(
-      get(urlMatching("/prisoners/.*/physical-attributes"))
+      get(urlMatching("/prisoners/$offenderNo/physical-attributes"))
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/prisonperson/PrisonPersonDpsApiMockServer.kt:71]: Stub individual prisoners for multi-prisoner tests

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/prisonperson/PrisonPersonReconService.kt (hunk 3 above)
Comment (steverendell) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/prisonperson/PrisonPersonReconService.kt:24]: You might want to consider the newer retry mechanism used by e.g. non-associations, it has an 'increasing delay' backoff strategy and more retries, see https://github.com/ministryofjustice/hmpps-prisoner-to-nomis-update/blob/d04d7922fa5b1fe5658baa8f8b9fe23e696cd399/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/nonassociations/NonAssociationsApiService.kt#L36

**Review by steverendell — APPROVED** [review]
lgtm with a comment

**Pull request merged**